File: src/mm_engine/order_book.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from enum import StrEnum
from typing import Deque


class Side(StrEnum):
    BUY = "buy"
    SELL = "sell"


class OrderType(StrEnum):
    LIMIT = "limit"
    MARKET = "market"


@dataclass
class Order:
    order_id: str
    side: Side
    order_type: OrderType
    quantity: int
    price: int | None
    sequence: int


@dataclass(frozen=True)
class Trade:
    symbol: str
    taker_order_id: str
    maker_order_id: str
    price: int
    quantity: int


class OrderBook:
    def __init__(self, symbol: str) -> None:
        self.symbol = symbol
        self.bids: dict[int, Deque[Order]] = {}
        self.asks: dict[int, Deque[Order]] = {}
        self.orders: dict[str, Order] = {}
        self.sequence = 0

    def add_limit_order(
        self,
        order_id: str,
        side: Side,
        price: int,
        quantity: int,
    ) -> list[Trade]:
        self._validate_new_order(order_id, quantity)
        order = self._new_order(order_id, side, OrderType.LIMIT, quantity, price)
        trades = self._match(order)

        if order.quantity > 0:
            self._rest(order)

        return trades

    def add_market_order(self, order_id: str, side: Side, quantity: int) -> list[Trade]:
        self._validate_new_order(order_id, quantity)
        order = self._new_order(order_id, side, OrderType.MARKET, quantity, None)
        return self._match(order)

    def cancel(self, order_id: str) -> bool:
        order = self.orders.pop(order_id, None)
        if order is None:
            return False

        levels = self.bids if order.side == Side.BUY else self.asks
        assert order.price is not None
        queue = levels.get(order.price)
        if queue is None:
            return False

        levels[order.price] = deque(resting for resting in queue if resting.order_id != order_id)
        if not levels[order.price]:
            del levels[order.price]
        return True

    def best_bid(self) -> int | None:
        return max(self.bids) if self.bids else None

    def best_ask(self) -> int | None:
        return min(self.asks) if self.asks else None

    def spread(self) -> int | None:
        bid = self.best_bid()
        ask = self.best_ask()
        if bid is None or ask is None:
            return None
        return ask - bid

    def depth(self, levels: int = 5) -> dict[str, list[dict[str, int]]]:
        return {
            "bids": self._depth_side(self.bids, reverse=True, levels=levels),
            "asks": self._depth_side(self.asks, reverse=False, levels=levels),
        }

    def _new_order(
        self,
        order_id: str,
        side: Side,
        order_type: OrderType,
        quantity: int,
        price: int | None,
    ) -> Order:
        self.sequence += 1
        return Order(
            order_id=order_id,
            side=side,
            order_type=order_type,
            quantity=quantity,
            price=price,
            sequence=self.sequence,
        )

    def _match(self, taker: Order) -> list[Trade]:
        trades: list[Trade] = []
        opposite = self.asks if taker.side == Side.BUY else self.bids

        while taker.quantity > 0 and opposite:
            best_price = min(opposite) if taker.side == Side.BUY else max(opposite)
            if not self._is_crossing(taker, best_price):
                break

            queue = opposite[best_price]
            while taker.quantity > 0 and queue:
                maker = queue[0]
                matched_quantity = min(taker.quantity, maker.quantity)
                taker.quantity -= matched_quantity
                maker.quantity -= matched_quantity
                trades.append(
                    Trade(
                        symbol=self.symbol,
                        taker_order_id=taker.order_id,
                        maker_order_id=maker.order_id,
                        price=best_price,
                        quantity=matched_quantity,
                    )
                )

                if maker.quantity == 0:
                    queue.popleft()
                    self.orders.pop(maker.order_id, None)

            if not queue:
                del opposite[best_price]

        return trades

    def _is_crossing(self, taker: Order, best_price: int) -> bool:
        if taker.order_type == OrderType.MARKET:
            return True
        assert taker.price is not None
        if taker.side == Side.BUY:
            return taker.price >= best_price
        return taker.price <= best_price

    def _rest(self, order: Order) -> None:
        assert order.price is not None
        levels = self.bids if order.side == Side.BUY else self.asks
        levels.setdefault(order.price, deque()).append(order)
        self.orders[order.order_id] = order

    def _depth_side(
        self,
        side: dict[int, Deque[Order]],
        reverse: bool,
        levels: int,
    ) -> list[dict[str, int]]:
        rows = []
        for price in sorted(side, reverse=reverse)[:levels]:
            rows.append(
                {
                    "price": price,
                    "quantity": sum(order.quantity for order in side[price]),
                    "orders": len(side[price]),
                }
            )
        return rows

    def _validate_new_order(self, order_id: str, quantity: int) -> None:
        if order_id in self.orders:
            raise ValueError(f"duplicate order_id: {order_id}")
        if quantity <= 0:
            raise ValueError("quantity must be positive")

File: native/src/order_book.cpp
#include "mm_engine/order_book.hpp"

#include <algorithm>
#include <stdexcept>
#include <utility>

namespace mm_engine {

OrderBook::OrderBook(std::string symbol) : symbol_(std::move(symbol)) {}

std::vector<Trade> OrderBook::add_limit_order(
    std::string order_id,
    Side side,
    Price price,
    Quantity quantity
) {
    validate_new_order(order_id, quantity);
    Order order{
        .order_id = std::move(order_id),
        .side = side,
        .quantity = quantity,
        .price = price,
        .sequence = ++sequence_,
    };
    auto trades = match(order, false);
    if (order.quantity > 0) {
        rest(std::move(order));
    }
    return trades;
}

std::vector<Trade> OrderBook::add_market_order(
    std::string order_id,
    Side side,
    Quantity quantity
) {
    validate_new_order(order_id, quantity);
    Order order{
        .order_id = std::move(order_id),
        .side = side,
        .quantity = quantity,
        .price = std::nullopt,
        .sequence = ++sequence_,
    };
    return match(order, true);
}

bool OrderBook::cancel(const std::string& order_id) {
    const auto location_it = resting_orders_.find(order_id);
    if (location_it == resting_orders_.end()) {
        return false;
    }

    const auto location = location_it->second;
    auto erase_from_level = [&order_id](auto& levels, Price price) {
        const auto level_it = levels.find(price);
        if (level_it == levels.end()) {
            throw std::logic_error("resting order index points to a missing price level");
        }

        auto& queue = level_it->second;
        const auto order_it = std::find_if(
            queue.begin(),
            queue.end(),
            [&order_id](const Order& order) { return order.order_id == order_id; }
        );
        if (order_it == queue.end()) {
            throw std::logic_error("resting order index points to a missing order");
        }

        queue.erase(order_it);
        if (queue.empty()) {
            levels.erase(level_it);
        }
    };

    if (location.side == Side::buy) {
        erase_from_level(bids_, location.price);
    } else {
        erase_from_level(asks_, location.price);
    }
    resting_orders_.erase(location_it);
    return true;
}

std::optional<Price> OrderBook::best_bid() const {
    if (bids_.empty()) {
        return std::nullopt;
    }
    return bids_.begin()->first;
}

std::optional<Price> OrderBook::best_ask() const {
    if (asks_.empty()) {
        return std::nullopt;
    }
    return asks_.begin()->first;
}

std::optional<Price> OrderBook::spread() const {
    const auto bid = best_bid();
    const auto ask = best_ask();
    if (!bid || !ask) {
        return std::nullopt;
    }
    return *ask - *bid;
}

std::vector<DepthLevel> OrderBook::bid_depth(std::size_t levels) const {
    std::vector<DepthLevel> result;
    result.reserve(std::min(levels, bids_.size()));
    for (const auto& [price, queue] : bids_) {
        Quantity quantity = 0;
        for (const auto& order : queue) {
            quantity += order.quantity;
        }
        result.push_back({price, quantity, queue.size()});
        if (result.size() == levels) {
            break;
        }
    }
    return result;
}

std::vector<DepthLevel> OrderBook::ask_depth(std::size_t levels) const {
    std::vector<DepthLevel> result;
    result.reserve(std::min(levels, asks_.size()));
    for (const auto& [price, queue] : asks_) {
        Quantity quantity = 0;
        for (const auto& order : queue) {
            quantity += order.quantity;
        }
        result.push_back({price, quantity, queue.size()});
        if (result.size() == levels) {
            break;
        }
    }
    return result;
}

std::size_t OrderBook::resting_order_count() const {
    return resting_orders_.size();
}

void OrderBook::validate_new_order(const std::string& order_id, Quantity quantity) const {
    if (resting_orders_.contains(order_id)) {
        throw std::invalid_argument("duplicate order_id: " + order_id);
    }
    if (quantity <= 0) {
        throw std::invalid_argument("quantity must be positive");
    }
}

std::vector<Trade> OrderBook::match(Order& taker, bool market_order) {
    if (taker.side == Side::buy) {
        return match_buy(taker, market_order);
    }
    return match_sell(taker, market_order);
}

std::vector<Trade> OrderBook::match_buy(Order& taker, bool market_order) {
    std::vector<Trade> trades;
    while (taker.quantity > 0 && !asks_.empty()) {
        auto level_it = asks_.begin();
        const auto best_price = level_it->first;
        if (!market_order && *taker.price < best_price) {
            break;
        }

        auto& queue = level_it->second;
        while (taker.quantity > 0 && !queue.empty()) {
            auto& maker = queue.front();
            const auto matched_quantity = std::min(taker.quantity, maker.quantity);
            taker.quantity -= matched_quantity;
            maker.quantity -= matched_quantity;
            trades.push_back({
                symbol_,
                taker.order_id,
                maker.order_id,
                best_price,
                matched_quantity,
            });

            if (maker.quantity == 0) {
                const auto maker_order_id = maker.order_id;
                queue.pop_front();
                resting_orders_.erase(maker_order_id);
            }
        }
        if (queue.empty()) {
            asks_.erase(level_it);
        }
    }
    return trades;
}

std::vector<Trade> OrderBook::match_sell(Order& taker, bool market_order) {
    std::vector<Trade> trades;
    while (taker.quantity > 0 && !bids_.empty()) {
        auto level_it = bids_.begin();
        const auto best_price = level_it->first;
        if (!market_order && *taker.price > best_price) {
            break;
        }

        auto& queue = level_it->second;
        while (taker.quantity > 0 && !queue.empty()) {
            auto& maker = queue.front();
            const auto matched_quantity = std::min(taker.quantity, maker.quantity);
            taker.quantity -= matched_quantity;
            maker.quantity -= matched_quantity;
            trades.push_back({
                symbol_,
                taker.order_id,
                maker.order_id,
                best_price,
                matched_quantity,
            });

            if (maker.quantity == 0) {
                const auto maker_order_id = maker.order_id;
                queue.pop_front();
                resting_orders_.erase(maker_order_id);
            }
        }
        if (queue.empty()) {
            bids_.erase(level_it);
        }
    }
    return trades;
}

void OrderBook::rest(Order order) {
    const auto price = *order.price;
    const auto order_id = order.order_id;
    const auto side = order.side;
    if (side == Side::buy) {
        bids_[price].push_back(std::move(order));
    } else {
        asks_[price].push_back(std::move(order));
    }
    resting_orders_.emplace(order_id, RestingLocation{side, price});
}

}  // namespace mm_engine
